benchmark/solver/solver_common.hpp

```diff
@@ -399,70 +370,141 @@ struct SolverGenerator : DefaultSystemGenerator<> {
 };
 
 
-template <typename VectorType>
-void solve_system(const std::string& solver_name,
-                  const std::string& precond_name,
-                  const char* precond_solver_name,
-                  std::shared_ptr<gko::Executor> exec,
-                  std::shared_ptr<Timer> timer,
-                  std::shared_ptr<const gko::LinOp> system_matrix,
-                  const VectorType* b, const VectorType* x,
-                  rapidjson::Value& test_case,
-                  rapidjson::MemoryPoolAllocator<>& allocator)
-{
-    try {
-        auto& solver_case = test_case["solver"];
-        if (!FLAGS_overwrite && solver_case.HasMember(precond_solver_name)) {
-            return;
+template <typename Generator>
+struct solver_benchmark_state {
+    using Vec = typename Generator::Vec;
+    std::shared_ptr<gko::LinOp> system_matrix;
+    std::unique_ptr<Vec> b;
+    std::unique_ptr<Vec> x;
+};
+
+
+template <typename Generator>
+struct SolverBenchmark : Benchmark<solver_benchmark_state<Generator>> {
+    std::string name;
+    std::vector<std::string> precond_solvers;
+    std::map<std::string, std::pair<std::string, std::string>> decoder;
+    Generator generator;
+
+    SolverBenchmark(Generator generator) : name{"solver"}, generator{generator}
+    {
+        auto solvers = split(FLAGS_solvers, ',');
+        auto preconds = split(FLAGS_preconditioners, ',');
+        for (const auto& s : solvers) {
+            for (const auto& p : preconds) {
+                precond_solvers.push_back(s + (p == "none" ? "" : "-" + p));
+                decoder[precond_solvers.back()] = {s, p};
+            }
+        }
+    }
+
+    const std::string& get_name() const override { return name; }
+
+    const std::vector<std::string>& get_operations() const override
+    {
+        return precond_solvers;
+    }
+
+    bool should_print() const override { return true; }
+
+    std::string get_example_config() const override
+    {
+        return solver_example_config;
+    }
+
+    bool validate_config(const json& value) const override
```

Comment:
I think this should call `generator.validate_config` same as in the SpMV benchmark. The check for `optimal` can be added to that.

Reply:
great suggestion!